File: python/day16.py
from adventofcode import AdventOfCode, Point, main
from collections import defaultdict


def CCW(vec):
    return vec @ [[0,-1],[-1,0]]

def CW(vec):
    return vec @ [[0, 1],[ 1,0]]

Reflection = {
    '/':  lambda vec: (CCW(vec),),
    '\\': lambda vec: (CW(vec),),
    '|':  lambda vec: (vec,) if vec.x == 0 else (CCW(vec), CW(vec)),
    '-':  lambda vec: (vec,) if vec.y == 0 else (CCW(vec), CW(vec)),
    '.':  lambda vec: (vec,)
    }


class Beams1(AdventOfCode):

    def __init__(self, inputfile):
        super().__init__(inputfile)
        self.rows = []
        self.width = 0
    
    @property
    def height(self):
        return len(self.rows)
    
    def line(self, line):
        line = line.strip()
        self.width = len(line)
        rows = ((col,prism) for col,prism in enumerate(line) if prism != '.')
        self.rows.append(defaultdict(lambda: '.', rows))

    def energize(self, start_at, start_dir):
        energized = set() # tuple(cel, vec)
        active = set()
        
        def passthru(_beam, _dir):
            if 0 <= _beam.row < self.height and 0 <= _beam.col < self.width:
                trace = (_beam, _dir)
                if trace not in energized:
                    energized.add(trace)
                    active.add(trace)
        
        passthru(start_at, start_dir)
        while active:
            beam,direction = active.pop()
            reflections = Reflection[self.rows[beam.row][beam.col]](direction)
            for bounce in reflections:
                passthru(beam+bounce, bounce)
            
        return len(set(tile for tile,vector in energized))

    def activate(self):
        return self.energize(Point(0,0), Point(1,0))


class Beams2(Beams1):

    def activate(self):
        most_energized = 0
        for j in range(self.height):
            energy = max(self.energize(Point(0,j), Point(1,0)),
                         self.energize(Point(self.width-1,j), Point(-1,0)))
            most_energized = max(most_energized, energy)
        for i in range(self.width):
            energy = max(self.energize(Point(i,0), Point(0,1)),
                         self.energize(Point(i,self.height-1), Point(0,-1)))
            most_energized = max(most_energized, energy)
        return most_energized


if __name__=='__main__':
    print(main([Beams1, Beams3]).read().activate())
    # for pt in [Point(1,0),Point(-1,0),Point(0,-1),Point(0,1)]:
    #     print(pt, CCW(pt), CW(pt))
    #     print(Reflection['/'](pt), Reflection['\\'](pt))
    #     for m in [ [[0,-1],[-1,0]], [[0,1],[1,0]] ]:
    #         print(str(m), pt@m)
    

File: lua/day16.lua
local aoc = require"adventofcode"
local Point,Grid = aoc.Point,aoc.Grid
local format,remove,extend = string.format,table.remove,aoc.extend

local Directions = {
    N = Point(0,-1);
    S = Point(0,1);
    W = Point(-1,0);
    E = Point(1,0);
}

local Reflections = {
    ['/'] = function (out)
                return {N={out.E}, S={out.W}, W={out.S}, E={out.N}}
            end;
   ['\\'] = function (out)
                return {N={out.W}, S={out.E}, W={out.N}, E={out.S}}
            end;
    ['|'] = function (out)
                local refl = {N={out.N}, S={out.S}, W={out.N,out.S}}
                refl.E = refl.W
                return refl
            end;
    ['-'] = function (out)
                local refl = {W={out.W}, E={out.E}, N={out.W,out.E}}
                refl.S = refl.N
                return refl
            end;
}

function analyze(grid)

    local function measure(start, dir)
        local steps = 0
        local stop = start
        local delta = Directions[dir]
        local mark = grid[stop+delta]
        while mark do
            steps = steps + 1
            stop = stop + delta
            if mark ~= '.' then
                break
            end
            mark = grid[stop+delta]
        end
        local beam = format("X%dY%d%s%d", start.x, start.y, dir, steps)
        return {beam, stop, dir}
    end

    local function make_beams(mirror, start)
        local out = {
            N = measure(start, 'N');
            S = measure(start, 'S');
            W = measure(start, 'W');
            E = measure(start, 'E');
        }
        return Reflections[mirror](out)
    end

    local beams = {}
    for j = 1,grid.height do
        beams[j] = {}
    end
    local found = grid:find("[/\\|-]")
    while found do
        local mirror = grid[found]
        beams[found.row][found.col] = make_beams(mirror, found)
        found = grid:next(found)
        if found then
            found = grid:find("[/\\|-]", found)
        end
    end

    local paths = {height=grid.height, width=grid.width}
    function trace(reflection, start)
        for i = #reflection,1,-1 do
            local beam = reflection[i]
            local key = beam[1]
            if not paths[key] then
                local stop = beam[2]
                if stop ~= start then
                    local dir = beam[3]
                    local mirror = beams[stop.j][stop.i]
                    if mirror then
                        bounce = mirror[dir]
                        paths[key] = {key, start, stop, bounce}
                        trace(bounce, stop)
                    else
                        paths[key] = {key, start, stop, {}}
                    end
                end
            end
            if not paths[key] then
                remove(reflection, i)
            else
                reflection[i] = paths[key]
            end
        end
    end

    for j = 1,grid.height do
        local reflection
        local entry = Point(1,j)
        local key = "Y"..j.."+"
        if not beams[j][1] then
            local beam = measure(entry, 'E')
            beam[1] = key
            reflection = {beam}
        else
            reflection = beams[j][1].E
            paths[key] = {key, entry, entry, reflection}
        end
        trace(reflection, entry)

        local key = "Y"..j.."-"
        entry = Point(grid.width,j)
        if not beams[j][grid.width] then
            local beam = measure(entry, 'W')
            beam[1] = key
            reflection = {beam}
        else
            reflection = beams[j][grid.width].W
            paths[key] = {key, entry, entry, reflection}
        end
        trace(reflection, entry)
    end
    for i = 1,grid.width do
        local reflection
        local entry = Point(i,1)
        local key = "X"..i.."+"
        if not beams[1][i] then
            local beam = measure(entry, 'S')
            beam[1] = key
            reflection = {beam}
        else
            reflection = beams[1][i].S
            paths[key] = {key, entry, entry, reflection}
        end
        trace(reflection, entry)

        local key = "X"..i.."-"
        entry = Point(i, grid.height)
        if not beams[grid.height][i] then
            local beam = measure(entry, 'N')
            beam[1] = key
            reflection = {beam}
        else
            reflection = beams[grid.height][i].N
            paths[key] = {key, entry, entry, reflection}
        end
        trace(reflection, entry)
    end

    return paths
end

function energize(paths, key)
    local points = {}
    local active = {}
    local beams = {}
    if not paths[key] then return nil end
    beams[1] = paths[key]
    while #beams > 0 do
        local beam = remove(beams)
        if not active[beam[1]] then
            extend(beams, beam[4])
            active[beam[1]] = true

            local start,stop = beam[2],beam[3]
            if start.col > stop.col then
                start,stop = stop,start
            end
            for i = start.col,stop.col do
                points[format("%d,%d", i, start.row)] = true
            end
            if start.row > stop.row then
                start,stop = stop,start
            end
            for j = start.row,stop.row do
                points[format("%d,%d", start.col, j)] = true
            end
        end
    end

    local energized = {}
    for key,_ in pairs(points) do
        energized[#energized+1] = key
    end
    return energized
end

-- main
local grid = Grid(io.input())
local paths = analyze(grid)
print(#energize(paths, "Y1+"))

local most = 0
for j=1,grid.height do
    local m = #energize(paths, 'Y'..j..'+')
    if m > most then most = m end
    m = #energize(paths, 'Y'..j..'-')
    if m > most then most = m end
end
for i=1,grid.width do
    local m = #energize(paths, 'X'..i..'+')
    if m > most then most = m end
    m = #energize(paths, 'X'..i..'-')
    if m > most then most = m end
end
print(most)
